### PR TITLE
FIX check the type of the entries in sys.modules

### PR DESCRIPTION
**Pull request by pierreglaser**
Some modules such as `coverage` can inject some non-module objects inside `sys.modules`, creating unpredictable bugs while trying to infer the module of a function. This currently affects one test inside the `PyPy` entry of `cloudpickle` CI. This PR adds an additional type check when manipulating entries of `sys.modules` during `_whichmodule`.

### CODE REVIEW

**Comment by pierreglaser** [comment]
Note that if we decide to move along with this change, the `try...catch` statement inside `whichmodule` is not tested anymore by our `test_faulty_module` (because `FaultyModule` won't pass the `isintance` check). To cover those specific lines, we would need to create a module-like object including a nested construct (like `MyMod.MyClass.my_method`, where `MyClass.__getattr__` raises an error) that we would try to pickle. It sounds like a very complicated test, but if @ogrisel you think it is worth it can write it.

File: cloudpickle/cloudpickle.py [hunk 1]
```diff
@@ -154,7 +154,13 @@ def _whichmodule(obj, name):
     # Protect the iteration by using a list copy of sys.modules against dynamic
     # modules that trigger imports of other modules upon calls to getattr.
     for module_name, module in list(sys.modules.items()):
-        if module_name == '__main__' or module is None:
+        # Some modules such as coverage can inject non-module objects inside
+        # sys.modules
+        if (
+                module_name == '__main__' or
+                module is None or
+                not isinstance(module, types.ModuleType)
```
Comment (pierreglaser) [inline cloudpickle/cloudpickle.py]: maybe we can do a `issubclass(type(module), types.ModuleType)` check instead? WDYT?

Comment (ogrisel) [reply cloudpickle/cloudpickle.py]: `isinstance(module, types.ModuleType)` is simpler and cleaner.

**Comment by ogrisel** [comment]
> Note that if we decide to move along with this change, the try...catch statement inside whichmodule is not tested anymore by our test_faulty_module (because FaultyModule won't pass the isintance check).

You mean the try / finally block? No I think we should keep it anyway. It's a good practice to have test not introduce any long lasting side effect. Let's keep the `test_faulty_module` as it is.

> To cover those specific lines, we would need to create a module-like object including a nested construct (like MyMod.MyClass.my_method, where MyClass.__getattr__ raises an error) that we would try to pickle. It sounds like a very complicated test, but if @ogrisel you think it is worth it can write it.

I think that would be useful to have a new deeper non-regression test for this indeed.

**Comment by pierreglaser** [comment]
> You mean the try / finally block?

No, I meant [these lines](https://github.com/cloudpipe/cloudpickle/pull/326/files#diff-d2a3618afedd4e124c532151eedbae09R166-R171).

**Comment by pierreglaser** [comment]
`test_non_module_object_passing_whichmodule` is a faithful reproducer of the original issue I spotted on `pypy` with `coverage`: a non-module object is inserted in `sys.modules` and miraculously passes a `whichmodule` check, but since it is not a module, `cloudpickle` errors out when carrying out routine checks on it such as `_is_dynamic`.

I also revamped `test_faulty_module` to include the coverage of the lines mentioned [here](https://github.com/cloudpipe/cloudpickle/pull/326#issuecomment-578087273).

File: tests/cloudpickle_test.py [hunk 2]
```diff
@@ -1048,35 +1048,85 @@ def __init__(self, x):
...
+            assert func_module_name != 'NonModuleObject'
+            assert func_module_name is None
+
+            depickled_func = pickle_depickle(func, protocol=self.protocol)
+            assert depickled_func(2) == 4
+
+        finally:
+            sys.modules.pop('NonModuleObject')
+
     def test_faulty_module(self):
```
Comment (ogrisel) [inline tests/cloudpickle_test.py]: I think this test could be renamed to `test_unrelated_faulty_module` and then add a comment such as:

> Check that pickling a dynamically defined function or class does not fail when introspecting the currently loaded modules in `sys.modules` as long as those faulty modules are unrelated to the class or function we are currently pickling.

File: tests/cloudpickle_test.py [hunk 3]
```diff
@@ -1048,35 +1048,85 @@ def __init__(self, x):
...
+            def __getattr__(self, name):
+                # We whitelist func so that a _whichmodule(func, None) call returns
+                # the NonModuleObject instance if a type check on the entries
+                # of sys.modules is not carried out, but manipulating this
+                # instance thinking it really is a module later on in the
+                # pickling process of func errors out
+                if name == 'func':
+                    return func
+                else:
+                    raise ValueError
```
Comment (pierreglaser) [inline tests/cloudpickle_test.py]:
```suggestion
                    raise AttributeError
```

File: tests/cloudpickle_test.py [hunk 4]
```diff
@@ -1048,35 +1048,85 @@ def __init__(self, x):
...
+                    raise AttributeError
+
+        non_module_object = NonModuleObject()
+
+        assert func(2) == 4
+        assert func is non_module_object.func
+
+        # Any manipulation of non_module_object relying on attribute access
+        # will raise an Exception
+        with pytest.raises(ValueError):
```
Comment (pierreglaser) [inline tests/cloudpickle_test.py]:
```suggestion
        with pytest.raises(AttributeError):
```

**Pull request merged**